lib/logstash/codecs/csv.rb

```diff
@@ -51,113 +56,119 @@ class LogStash::Codecs::CSV < LogStash::Codecs::Base
   # "CP1252".
   config :charset, :validate => ::Encoding.name_list, :default => "UTF-8"
 
+  CONVERTERS = {
+    :integer => lambda do |value|
+      CSV::Converters[:integer].call(value)
+    end,
+
+    :float => lambda do |value|
+      CSV::Converters[:float].call(value)
+    end,
+
+    :date => lambda do |value|
+      result = CSV::Converters[:date].call(value)
+      result.is_a?(Date) ? LogStash::Timestamp.new(result.to_time) : result
+    end,
+
+    :date_time => lambda do |value|
+      result = CSV::Converters[:date_time].call(value)
+      result.is_a?(DateTime) ? LogStash::Timestamp.new(result.to_time) : result
+    end,
+
+    :boolean => lambda do |value|
+       value = value.strip.downcase
+       return false if value == "false"
+       return true  if value == "true"
+       return value
+    end
+  }
+  CONVERTERS.default = lambda {|v| v}
+  CONVERTERS.freeze
+
   def register
     @converter = LogStash::Util::Charset.new(@charset)
     @converter.logger = @logger
 
     # validate conversion types to be the valid ones.
-    @convert.each_pair do |column, type|
-      if !VALID_CONVERT_TYPES.include?(type)
-        raise LogStash::ConfigurationError, "#{type} is not a valid conversion type."
-      end
+    bad_types = @convert.values.select do |type|
+      !CONVERTERS.has_key?(type.to_sym)
+    end.uniq
+    raise(LogStash::ConfigurationError, "Invalid conversion types: #{bad_types.join(', ')}") unless bad_types.empty?
+
+    # @convert_symbols contains the symbolized types to avoid symbol conversion in the transform method
+    @convert_symbols = @convert.inject({}){|result, (k, v)| result[k] = v.to_sym; result}
+
+    # if the zero byte character is entered in the config, set the value
+    if (@quote_char == "\\x00")
+      @quote_char = "\x00"
     end
 
-    @headers = false
-    @options = { :col_sep => @separator, :quote_char => @quote_char }
+    @logger.debug? && @logger.debug("CSV parsing options", :col_sep => @separator, :quote_char => @quote_char)
   end
 
   def decode(data)
     data = @converter.convert(data)
     begin
-      values = CSV.parse_line(data, @options)
-      if @include_headers && !@headers
-        @headers = true
-        @options[:headers] = values
-      else
-        decoded = {}
-        values.each_with_index do |fields, index|
-          field_name, value = nil, nil
-          if fields.is_a?(String) && !( @skip_empty_columns && fields.nil?)  # No headers
-            next if ignore_field?(index)
-            field_name =  ( !@columns[index].nil? ? @columns[index] : "column#{(index+1)}")
-            value      = fields
-          elsif fields.is_a?(Array) # Got headers
-            field_name = fields[0]
-            value      = fields[1]
+      values = CSV.parse_line(data, :col_sep => @separator, :quote_char => @quote_char)
+
+      if (@autodetect_column_names && @columns.empty?)
+        @columns = values
```

Comment:
nitpick: would it be worth logging that we have detected the columns?

Reply:
we don't in the csv filter but I guess we could ... at debug level?